src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsEntityTypeDefinition.cs

```diff
@@ -112,7 +112,30 @@ private IEnumerable<IEdmStructuralProperty> ComputeDeclaredKey()
                 List<IEdmStructuralProperty> key = new List<IEdmStructuralProperty>();
                 foreach (CsdlPropertyReference keyProperty in this.entity.Key.Properties)
                 {
-                    IEdmStructuralProperty structuralProperty = this.FindProperty(keyProperty.PropertyName) as IEdmStructuralProperty;
+                    IEdmStructuralProperty structuralProperty = null;
+                    if (!string.IsNullOrEmpty(keyProperty.Alias))
+                    {
+                        string[] keyProperties = keyProperty.PropertyName.Split('/');
+                        for (int i = 0; i < keyProperties.Length; i++)
+                        {
+                            structuralProperty = this.FindProperty(keyProperties[i]) as IEdmStructuralProperty;
+
+                            if (structuralProperty.Type.Definition.TypeKind == EdmTypeKind.Complex)
+                            {
+                                structuralProperty = structuralProperty.Type.AsComplex().FindProperty(keyProperties[i]) as IEdmStructuralProperty;

```

Comment:
After looking at the tests, I think I know the tests are passing despite this apparent bug:
- In your tests you have an entity type `Product` which has a key property that aliases `ExampleObject/Name`
- `Product` also has a `Name` property
- When this method is called, the key property name is split into `[ExampleObject, Name]`
- In the first iteration (`i = 0`), `structuralProperty = this.FindProperty(keyProperties[i])` sets the `structureProperty` to `ExampleObject` initially.
- But then in the same iteration you call `ExampleObject.Type...FindProperty(keyProerties[i])` which attempts to find `ExampleObject` but returns null cause the property doesn't exist
- In the second iteration `i = 1`, you replace `structuralProperty = this.FindProperty(keyProperties[i])`, in this case the property path segment is `Name`. Since `Name` is also a property of `Product`, `Product.Name` is returned (not ExampleObject.Name), since it's neither a complex or entity type, the if statements are skipped over, avoiding the `NullReferenceException`
- So this actual returns the wrong property for the key, but since the test doesn't assert the type, declaring type, etc. of the returned property, it doesn't detect this bug.

Reply:
> I initially assumed this loop is trying to walk down the property path to find the actual property being referenced. So let's assume `PropertyName` is initially `Address/City/Town`, and `keyProperties` become `[Address, City, Town]`.
> 
> When `i` = 0:
> 
> * `structuralProperty = this.FindProperty(keyProperties[i])` will set the `structuralProperty` to `Address`
> * then if `structuralProperty` is a complex type, then wouldn't `structuralProperty...FindProperty(keyProperties[i])` also try to find the `Address` property in `structuralProperty`? Isn't that an error? Shouldn't we be trying to find the property `City` instead which is `keyProperties[i + 1]`?. What does `Address.FindProperty(Address)` actually return?
> 
> If trying to find a property that doesn't exist returns `null`, then next iteration of the loop will throw a `NullReferenceException`.
> 
> Is this method covered in tests? I assume it should cause tests to fail? Or I am misunderstanding this code.

Yes I agree with this.